cmd/ccboc/cmd/commands.go

```diff
@@ -0,0 +1,278 @@
+	StatusCode int    `json:"status_code"`
+}
+
+var globalConfig *config.Config
+
+func initializeConfig() error {
+	configPath, err := config.GetDefaultConfigPath()
+	if err != nil {
+		logrus.WithError(err).Fatal("couldn't get configuration path")
+	}
+
+	globalConfig, err = config.LoadConfig(configPath)
+	if err != nil {
+		logrus.WithError(err).Fatal("coudln't load the configuration file")
+	}
+	return nil
+}
+
+func login() error {
+	if token == "" || url == "" {
+		return fmt.Errorf("--token and --url must specified together")
+	}
+
+	cfg := config.Config{
+		APIURL: url,
+		Token:  token,
+	}
+
+	if err := cfg.Validate(); err != nil {
+		return fmt.Errorf("config validation failed: %v", err)
+	}
+
+	user, err := user.Current()
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	home := user.HomeDir
+
+	data, err := json.MarshalIndent(cfg, "", "")
+	if err != nil {
+		return fmt.Errorf("couldn't marthal the configuration file: %v", err)
+	}
+
+	path := filepath.Join(home, config.DefaultConfigPath)
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		if err := os.Mkdir(path, os.ModePerm); err != nil {
+			return fmt.Errorf("couldn't create dir: %v", err)
+		}
+	}
+
+	configFilePath := filepath.Join(home, config.DefaultConfigPath, config.DefaultConfigFileName)
+	if err := ioutil.WriteFile(configFilePath, data, 0777); err != nil {
+		return fmt.Errorf("couldn't write to file: %v", err)
+	}
+
+	fmt.Printf("Configuration file generated at %s\n", configFilePath)
+	return nil
+}
+
+func getCalculations() error {
+	body, responseError, err := request("GET", globalConfig.APIURL+"/calculations", bytes.NewBuffer(nil))
+	if err != nil {
+		logrus.WithError(err).Fatal("error while perfoming the http request")
+	} else if responseError != nil {
+		logrus.WithFields(logrus.Fields{"message": responseError.Message, "status_code": responseError.StatusCode}).Fatal("errors occurred")
+	}
+
+	var data *calculationsv1.CalculationList
+	if err := json.Unmarshal(body, &data); err != nil {
+		logrus.WithError(err).Fatal("Error unmarshaling json response")
+	}
+	output(data)
+	return nil
+}
+
+func getCalculationID() error {
+	args := os.Args
+	calcID := args[len(args)-1]
+
+	body, responseError, err := request("GET", globalConfig.APIURL+"/calculation/"+calcID, bytes.NewBuffer(nil))
+	if err != nil {
+		logrus.WithError(err).Fatal("error while perfoming the http request")
+	} else if responseError != nil {
+		logrus.WithFields(logrus.Fields{"message": responseError.Message, "status_code": responseError.StatusCode}).Fatal("errors occurred")
+	}
+
+	var calc *calculationsv1.Calculation
+	if err := json.Unmarshal(body, &calc); err != nil {
+		logrus.WithError(err).Fatal("Error unmarshaling json response")
+	}
+
+	output(calc)
+	return nil
+}
+
+func getCalculationResult() error {
+	req, err := http.NewRequest(http.MethodGet, globalConfig.APIURL+"/calculations/results?teff="+fmt.Sprintf("%0.1f", teff)+"&logg="+fmt.Sprintf("%0.2f", logG), bytes.NewBuffer(nil))
```

Comment:
since you are doing that, can you fix also this? We should construct the URL with the params, instead of just passing it as a constructed string here.

Reply:
I am thinking of deleting the whole struct `Config` which holds token and URL. Then I have to change the function `Validate()` to take a string instead of being attached to the struct.

Reply:
No, I mean to improve the way you construct the URL here.

Reply:
Like `url := globalConfig.APIURL+"/calculations/results?teff="+fmt.Sprintf("%0.1f", teff)+"&logg="+fmt.Sprintf("%0.2f", logG)` and then pass `url` to the function `http.NewRequest(...)`?

Reply:
```go
	u, err := url.Parse(globalConfig.APIURL)
	if err != nil {
		...
	}

	u.Path = "/calculations/results"

	q := u.Query()
	q.Set("teff", fmt.Sprintf("%0.1f", teff))
	q.Set("logG", fmt.Sprintf("%0.2f", logG))
	u.RawQuery = q.Encode()

 	req, err := http.NewRequest(http.MethodGet, u.String(), bytes.NewBuffer(nil))
....
....
```

Perhaps we can do the parsing while we create the global config and make the `globalConfig.APIURL` a `url.URL` instead of a string, and then use it everywhere.